File: app/src/core/qbnn.ts
/**
 * QBNN -- APQB neural network layer (paper v2 Sec. 4-5, Eq. 19-32).
 *   r = tanh(h); a = W h + b; q = tanh(a)
 *   g_j = sum_{S, 1<=|S|<=K} J_{S,j} prod_{i in S} r_i
 *   Delta = q (.) g;  a~ = a + lambda Delta;  h' = sigma(a~)
 */
import { cexp, cmul, Complex } from './complex';
import { Rng } from './rng';

export type Subset = number[];

function combinations(n: number, k: number): Subset[] {
  const out: Subset[] = [];
  const rec = (start: number, cur: number[]) => {
    if (cur.length === k) {
      out.push([...cur]);
      return;
    }
    for (let i = start; i < n; i++) {
      cur.push(i);
      rec(i + 1, cur);
      cur.pop();
    }
  };
  rec(0, []);
  return out;
}

export const subsetKey = (S: Subset): string => (S.length ? S.join(',') : '∅');

/** phi_S(r) = prod_{i in S} r_i for |S| <= K (Eq. 19). */
export function subsetFeatures(r: number[], K?: number): Map<string, number> {
  const n = r.length;
  const kk = K ?? n;
  const feats = new Map<string, number>([['∅', 1]]);
  for (let k = 1; k <= kk; k++) for (const S of combinations(n, k)) feats.set(subsetKey(S), S.reduce((p, i) => p * r[i], 1));
  return feats;
}

/** Phi_S(z) = prod z_i, z_i = exp(i 2 theta_i) (Eq. 21-22). */
export function complexSubsetFeatures(thetas: number[], K?: number): Map<string, Complex> {
  const z = thetas.map((t) => cexp(2 * t));
  const kk = K ?? z.length;
  const feats = new Map<string, Complex>([['∅', { re: 1, im: 0 }]]);
  for (let k = 1; k <= kk; k++) for (const S of combinations(z.length, k)) feats.set(subsetKey(S), S.reduce((p, i) => cmul(p, z[i]), { re: 1, im: 0 }));
  return feats;
}

/** eta = sqrt(1 - r^2 + eps) (Eq. 31). */
export const uncertainty = (r: number, eps = 1e-6): number => Math.sqrt(Math.max(1 - r * r, 0) + eps);
/** Monotone map eta -> [lo, hi] (Eq. 11 / 32). */
export const controlSignal = (eta: number, lo: number, hi: number): number => lo + (hi - lo) * eta;

export type ActivationName = 'tanh' | 'identity' | 'sigmoid' | 'relu';
const ACT: Record<ActivationName, (x: number) => number> = {
  tanh: Math.tanh,
  identity: (x) => x,
  sigmoid: (x) => 1 / (1 + Math.exp(-x)),
  relu: (x) => (x > 0 ? x : 0),
};

export interface LayerJSON {
  in_dim: number;
  out_dim: number;
  K: number;
  lam: number;
  activation: ActivationName;
  W: number[][];
  b: number[];
  J: number[][];
}

export type ParamRef = { row: number[]; i: number };

export class QBNNLayer {
  inDim: number;
  outDim: number;
  K: number;
  lam: number;
  activation: ActivationName;
  W: number[][];
  b: number[];
  subsets: Subset[] = [];
  J: number[][];
  last: { r: number[]; a: number[]; q: number[]; g: number[]; delta: number[]; eta: number[] } | null = null;

  constructor(inDim: number, outDim: number, K = 1, lam = 1, activation: ActivationName = 'tanh', rng?: Rng, initScale = 0.5) {
    const r = rng ?? new Rng(0);
    this.inDim = inDim;
    this.outDim = outDim;
    this.K = Math.max(0, Math.min(K, inDim));
    this.lam = lam;
    this.activation = activation;
    this.W = Array.from({ length: outDim }, () => Array.from({ length: inDim }, () => (r.random() * 2 - 1) * initScale));
    this.b = new Array(outDim).fill(0);
    for (let k = 1; k <= this.K; k++) this.subsets.push(...combinations(inDim, k));
    this.J = this.subsets.map(() => new Array(outDim).fill(0));
  }

  parameters(): ParamRef[] {
    const refs: ParamRef[] = [];
    for (const row of this.W) for (let i = 0; i < row.length; i++) refs.push({ row, i });
    for (let i = 0; i < this.b.length; i++) refs.push({ row: this.b, i });
    for (const row of this.J) for (let i = 0; i < row.length; i++) refs.push({ row, i });
    return refs;
  }

  numParameters(): number {
    return this.outDim * this.inDim + this.outDim + this.subsets.length * this.outDim;
  }

  forward(h: number[]): number[] {
    const r = h.map(Math.tanh);
    const a = this.W.map((row, j) => row.reduce((s, w, i) => s + w * h[i], 0) + this.b[j]);
    const q = a.map(Math.tanh);
    const g = new Array(this.outDim).fill(0);
    if (this.lam !== 0 && this.subsets.length) {
      this.subsets.forEach((S, s) => {
        const phi = S.reduce((p, i) => p * r[i], 1);
        if (phi === 0) return;
        const Jrow = this.J[s];
        for (let j = 0; j < this.outDim; j++) g[j] += Jrow[j] * phi;
      });
    }
    const delta = q.map((qq, j) => qq * g[j]);
    const aTilde = a.map((aa, j) => aa + this.lam * delta[j]);
    const act = ACT[this.activation];
    this.last = { r, a, q, g, delta, eta: r.map((v) => uncertainty(v)) };
    return aTilde.map(act);
  }

  toJSON(): LayerJSON {
    return { in_dim: this.inDim, out_dim: this.outDim, K: this.K, lam: this.lam, activation: this.activation, W: this.W, b: this.b, J: this.J };
  }

  static fromJSON(d: LayerJSON): QBNNLayer {
    const l = new QBNNLayer(d.in_dim, d.out_dim, d.K ?? 1, d.lam ?? 1, d.activation ?? 'tanh');
    l.W = d.W.map((row) => row.map(Number));
    l.b = d.b.map(Number);
    l.J = d.J.map((row) => row.map(Number));
    return l;
  }
}

export interface QBNNJSON {
  dims: number[];
  layers: LayerJSON[];
}

export class QBNN {
  dims: number[];
  layers: QBNNLayer[] = [];

  constructor(dims: number[], K = 1, lam = 1, activation: ActivationName = 'tanh', outputActivation: ActivationName = 'identity', seed = 0) {
    const rng = new Rng(seed);
    this.dims = [...dims];
    for (let i = 0; i < dims.length - 1; i++) {
      const last = i === dims.length - 2;
      this.layers.push(new QBNNLayer(dims[i], dims[i + 1], K, lam, last ? outputActivation : activation, rng));
    }
  }

  forward(x: number[]): number[] {
    let h = [...x];
    for (const l of this.layers) h = l.forward(h);
    return h;
  }

  parameters(): ParamRef[] {
    return this.layers.flatMap((l) => l.parameters());
  }

  numParameters(): number {
    return this.layers.reduce((s, l) => s + l.numParameters(), 0);
  }

  uncertainties(): number[][] {
    return this.layers.map((l) => l.last?.eta ?? []);
  }

  toJSON(): QBNNJSON {
    return { dims: this.dims, layers: this.layers.map((l) => l.toJSON()) };
  }

  static fromJSON(d: QBNNJSON): QBNN {
    const net = new QBNN(d.dims);
    net.layers = d.layers.map((ld) => QBNNLayer.fromJSON(ld));
    return net;
  }
}

export type Sample = [number[], number[]];

export const xorDataset = (): Sample[] => [[[-1, -1], [-1]], [[-1, 1], [1]], [[1, -1], [1]], [[1, 1], [-1]]];

export function parityDataset(n: number): Sample[] {
  const data: Sample[] = [];
  for (let idx = 0; idx < 2 ** n; idx++) {
    const x: number[] = [...Array(n).keys()].map((i) => ((idx >> i) & 1 ? 1 : -1));
    data.push([x, [x.reduce((p, v) => p * v, 1)]]);
  }
  return data;
}

export function mse(net: QBNN, data: Sample[]): number {
  let total = 0;
  for (const [x, y] of data) {
    const out = net.forward(x);
    total += out.reduce((s, o, i) => s + (o - y[i]) ** 2, 0);
  }
  return total / data.length;
}

export function accuracy(net: QBNN, data: Sample[]): number {
  let ok = 0;
  for (const [x, y] of data) {
    const out = net.forward(x);
    if (out.every((o, i) => o > 0 === y[i] > 0)) ok++;
  }
  return ok / data.length;
}

export interface TrainOptions {
  epochs?: number;
  lr?: number;
  eps?: number;
  log?: (line: string) => void;
  logEvery?: number;
  onEpoch?: (epoch: number, loss: number) => void;
}

/** Gradient descent with central finite differences (small hypothesis-testing nets). */
export function train(net: QBNN, data: Sample[], opts: TrainOptions = {}): number[] {
  const epochs = opts.epochs ?? 200;
  const lr = opts.lr ?? 0.1;
  const eps = opts.eps ?? 1e-4;
  const logEvery = opts.logEvery ?? 50;
  const history: number[] = [];
  const params = net.parameters();
  for (let epoch = 1; epoch <= epochs; epoch++) {
    const grads: number[] = [];
    for (const { row, i } of params) {
      const orig = row[i];
      row[i] = orig + eps;
      const lp = mse(net, data);
      row[i] = orig - eps;
      const lm = mse(net, data);
      row[i] = orig;
      grads.push((lp - lm) / (2 * eps));
    }
    params.forEach(({ row, i }, k) => (row[i] -= lr * grads[k]));
    const loss = mse(net, data);
    history.push(loss);
    opts.onEpoch?.(epoch, loss);
    if (opts.log && (epoch % logEvery === 0 || epoch === 1 || epoch === epochs)) {
      opts.log(`epoch ${String(epoch).padStart(4)}  loss ${loss.toFixed(6)}  acc ${accuracy(net, data).toFixed(2)}`);
    }
  }
  return history;
}

/** Train one epoch (for incremental UI progress). */
export function trainStep(net: QBNN, data: Sample[], lr = 0.1, eps = 1e-4): number {
  const params = net.parameters();
  const grads = params.map(({ row, i }) => {
    const orig = row[i];
    row[i] = orig + eps;
    const lp = mse(net, data);
    row[i] = orig - eps;
    const lm = mse(net, data);
    row[i] = orig;
    return (lp - lm) / (2 * eps);
  });
  params.forEach(({ row, i }, k) => (row[i] -= lr * grads[k]));
  return mse(net, data);
}

File: qubit_computer/qbnn.py
"""QBNN -- APQB neural network layer (revised manuscript v2, Sec. 4-5).

Pure-Python, dependency-free implementation of:

* subset-product features  phi_S(r) = prod_{i in S} r_i           (Eq. 19)
* complex subset features   Phi_S(z) = prod_{i in S} z_i           (Eq. 22)
* the basic QBNN layer                                             (Eq. 23-29)

      r      = tanh(h)                       bounded APQB coordinates
      a      = W h + b                       classical pre-activation
      q      = tanh(a)                       candidate state
      g_j    = sum_{S, 1<=|S|<=K} J_{S,j} prod_{i in S} r_i   (Eq. 26 / 30)
      Delta  = q (.) g                       multiplicative gate
      a~     = a + lambda * Delta
      h_next = sigma(a~)

* APQB uncertainty amplitude eta_i = sqrt(1 - r_i^2 + eps) mapped to a
  control signal (dropout rate / temperature)                      (Eq. 31-32)

With ``lam = 0`` (or J = 0) the layer is an ordinary affine + activation
layer.  Training uses finite-difference gradients, which is exact enough
for the small hypothesis-testing tasks (XOR, parity) the paper proposes
in Sec. 6 and keeps the OS free of third-party dependencies.
"""

from __future__ import annotations

import cmath
import math
import random
from itertools import combinations
from typing import Callable, Dict, List, Optional, Sequence, Tuple

__all__ = [
    "subset_features", "complex_subset_features", "uncertainty", "control_signal",
    "QBNNLayer", "QBNN", "train", "xor_dataset", "parity_dataset",
]


def subset_features(r: Sequence[float], K: Optional[int] = None) -> Dict[Tuple[int, ...], float]:
    """phi_S(r) for every S with |S| <= K (Eq. 19).  phi_{} = 1."""
    n = len(r)
    if K is None:
        K = n
    feats: Dict[Tuple[int, ...], float] = {(): 1.0}
    for k in range(1, K + 1):
        for S in combinations(range(n), k):
            prod = 1.0
            for i in S:
                prod *= r[i]
            feats[S] = prod
    return feats


def complex_subset_features(thetas: Sequence[float], K: Optional[int] = None) -> Dict[Tuple[int, ...], complex]:
    """Phi_S(z) with z_i = exp(i 2 theta_i) (Eq. 21-22)."""
    z = [cmath.exp(1j * 2.0 * t) for t in thetas]
    n = len(z)
    if K is None:
        K = n
    feats: Dict[Tuple[int, ...], complex] = {(): 1 + 0j}
    for k in range(1, K + 1):
        for S in combinations(range(n), k):
            prod = 1 + 0j
            for i in S:
                prod *= z[i]
            feats[S] = prod
    return feats


def uncertainty(r: float, eps: float = 1e-6) -> float:
    """eta = sqrt(1 - r^2 + eps) (Eq. 31)."""
    return math.sqrt(max(1.0 - r * r, 0.0) + eps)


def control_signal(eta: float, lo: float, hi: float) -> float:
    """Monotone map eta -> [lo, hi], e.g. temperature (Eq. 11) or dropout (Eq. 32)."""
    if not all(math.isfinite(v) for v in (eta, lo, hi)) or lo > hi:
        raise ValueError("invalid control signal bounds")
    # Preserve Eq. 31; saturate its epsilon overshoot at the control boundary.
    t = max(0.0, min(1.0, eta))
    return max(lo, min(hi, (1.0 - t) * lo + t * hi))


def _tanh(x: float) -> float:
    return math.tanh(x)


def _identity(x: float) -> float:
    return x


ACTIVATIONS: Dict[str, Callable[[float], float]] = {
    "tanh": _tanh,
    "identity": _identity,
    "sigmoid": lambda x: 1.0 / (1.0 + math.exp(-x)) if x > -700 else 0.0,
    "relu": lambda x: x if x > 0 else 0.0,
}


class QBNNLayer:
    """One QBNN layer, Eq. (23)-(29) with K-order gating (Eq. 30)."""

    def __init__(self, in_dim: int, out_dim: int, K: int = 1, lam: float = 1.0,
                 activation: str = "tanh", rng: Optional[random.Random] = None,
                 init_scale: float = 0.5):
        rng = rng or random.Random(0)
        self.in_dim = in_dim
        self.out_dim = out_dim
        self.K = max(0, min(K, in_dim))
        self.lam = lam
        self.activation_name = activation
        self.act = ACTIVATIONS[activation]
        self.W = [[rng.uniform(-init_scale, init_scale) for _ in range(in_dim)] for _ in range(out_dim)]
        self.b = [0.0] * out_dim
        # Subsets S with 1 <= |S| <= K, in a fixed order (Eq. 30).
        self.subsets: List[Tuple[int, ...]] = []
        for k in range(1, self.K + 1):
            self.subsets.extend(combinations(range(in_dim), k))
        # J[s][j] for subset index s, output unit j; zero-init keeps the plain layer.
        self.J = [[0.0] * out_dim for _ in self.subsets]
        self.last: Dict[str, List[float]] = {}

    # ------------------------------------------------------------ params
    def parameters(self) -> List[Tuple[List[float], int]]:
        """Flat list of (row, index) handles for in-place updates."""
        handles: List[Tuple[List[float], int]] = []
        for row in self.W:
            handles.extend((row, i) for i in range(len(row)))
        handles.extend((self.b, i) for i in range(len(self.b)))
        for row in self.J:
            handles.extend((row, i) for i in range(len(row)))
        return handles

    def num_parameters(self) -> int:
        return self.out_dim * self.in_dim + self.out_dim + len(self.subsets) * self.out_dim

    # ----------------------------------------------------------- forward
    def forward(self, h: Sequence[float]) -> List[float]:
        if len(h) != self.in_dim or not all(math.isfinite(x) for x in h):
            raise ValueError(f"expected {self.in_dim} finite inputs")
        r = [math.tanh(x) for x in h]                                   # (23)
        a = [sum(w * x for w, x in zip(row, h)) + bb for row, bb in zip(self.W, self.b)]  # (24)
        q = [math.tanh(x) for x in a]                                   # (25)
        g = [0.0] * self.out_dim                                        # (26)/(30)
        if self.lam != 0.0 and self.subsets:
            for s, S in enumerate(self.subsets):
                phi = 1.0
                for i in S:
                    phi *= r[i]
                if phi == 0.0:
                    continue
                Jrow = self.J[s]
                for j in range(self.out_dim):
                    g[j] += Jrow[j] * phi
        delta = [qq * gg for qq, gg in zip(q, g)]                       # (27)
        a_tilde = [aa + self.lam * dd for aa, dd in zip(a, delta)]      # (28)
        out = [self.act(x) for x in a_tilde]                            # (29)
        self.last = {"r": r, "a": a, "q": q, "g": g, "delta": delta, "eta": [uncertainty(x) for x in r]}
        return out

    __call__ = forward

    def to_dict(self) -> dict:
        return {
            "in_dim": self.in_dim, "out_dim": self.out_dim, "K": self.K, "lam": self.lam,
            "activation": self.activation_name, "W": self.W, "b": self.b, "J": self.J,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "QBNNLayer":
        layer = cls(d["in_dim"], d["out_dim"], K=d.get("K", 1), lam=d.get("lam", 1.0),
                    activation=d.get("activation", "tanh"))
        layer.W = [list(map(float, row)) for row in d["W"]]
        layer.b = list(map(float, d["b"]))
        layer.J = [list(map(float, row)) for row in d["J"]]
        return layer


class QBNN:
    """A stack of QBNN layers."""

    def __init__(self, dims: Sequence[int], K: int = 1, lam: float = 1.0,
                 activation: str = "tanh", output_activation: str = "identity",
                 seed: int = 0):
        rng = random.Random(seed)
        self.dims = list(dims)
        self.layers: List[QBNNLayer] = []
        for i in range(len(dims) - 1):
            last = i == len(dims) - 2
            self.layers.append(QBNNLayer(dims[i], dims[i + 1], K=K, lam=lam,
                                         activation=output_activation if last else activation,
                                         rng=rng))

    def forward(self, x: Sequence[float]) -> List[float]:
        h = list(x)
        for layer in self.layers:
            h = layer.forward(h)
        return h

    __call__ = forward

    def parameters(self) -> List[Tuple[List[float], int]]:
        out: List[Tuple[List[float], int]] = []
        for layer in self.layers:
            out.extend(layer.parameters())
        return out

    def num_parameters(self) -> int:
        return sum(l.num_parameters() for l in self.layers)

    def uncertainties(self) -> List[List[float]]:
        """eta of every unit in every layer after the last forward pass (Eq. 31)."""
        return [layer.last.get("eta", []) for layer in self.layers]

    def to_dict(self) -> dict:
        return {"dims": self.dims, "layers": [l.to_dict() for l in self.layers]}

    @classmethod
    def from_dict(cls, d: dict) -> "QBNN":
        net = cls(d["dims"])
        net.layers = [QBNNLayer.from_dict(ld) for ld in d["layers"]]
        return net


# ------------------------------------------------------------------ data
def xor_dataset() -> List[Tuple[List[float], List[float]]]:
    return [([-1.0, -1.0], [-1.0]), ([-1.0, 1.0], [1.0]), ([1.0, -1.0], [1.0]), ([1.0, 1.0], [-1.0])]


def parity_dataset(n: int) -> List[Tuple[List[float], List[float]]]:
    """n-bit parity with +-1 encoding: target = prod_i x_i (a top-order subset feature)."""
    data = []
    for idx in range(2 ** n):
        x = [1.0 if (idx >> i) & 1 else -1.0 for i in range(n)]
        y = 1.0
        for v in x:
            y *= v
        data.append((x, [y]))
    return data


def mse(net: QBNN, data: Sequence[Tuple[Sequence[float], Sequence[float]]]) -> float:
    total = 0.0
    for x, y in data:
        out = net.forward(x)
        total += sum((o - t) ** 2 for o, t in zip(out, y))
    return total / len(data)


def accuracy(net: QBNN, data: Sequence[Tuple[Sequence[float], Sequence[float]]]) -> float:
    ok = 0
    for x, y in data:
        out = net.forward(x)
        ok += all((o > 0) == (t > 0) for o, t in zip(out, y))
    return ok / len(data)


def train(net: QBNN, data: Sequence[Tuple[Sequence[float], Sequence[float]]],
          epochs: int = 200, lr: float = 0.1, eps: float = 1e-4,
          log: Optional[Callable[[str], None]] = None, log_every: int = 50) -> List[float]:
    """Gradient descent with central finite differences (exact enough for tiny nets)."""
    history: List[float] = []
    params = net.parameters()
    for epoch in range(1, epochs + 1):
        grads: List[float] = []
        for row, i in params:
            orig = row[i]
            row[i] = orig + eps
            lp = mse(net, data)
            row[i] = orig - eps
            lm = mse(net, data)
            row[i] = orig
            grads.append((lp - lm) / (2 * eps))
        for (row, i), g in zip(params, grads):
            row[i] -= lr * g
        loss = mse(net, data)
        history.append(loss)
        if log and (epoch % log_every == 0 or epoch == 1 or epoch == epochs):
            log(f"epoch {epoch:4d}  loss {loss:.6f}  acc {accuracy(net, data):.2f}")
    return history
